### PR TITLE
OCPBUGS-35099: OCPBUGS-4466: Prevent cluster installation with mismatched worker assets and worker replicas

### PR DESCRIPTION
**Pull request by sadasu**
This is updating the fix introduced in https://github.com/openshift/installer/pull/8226.

Based on https://docs.openshift.com/container-platform/4.13/installing/installing_gcp/installing-gcp-three-node.html, compact clusters on GCP are actually supported so the fix has been updated to disallow only incorrect states that we may end up in with incorrect manipulation of generated assets as detailed in the big.

Also, limiting the fix to just the GCP platform.

### CODE REVIEW

**Comment by sadasu** [comment]
With the steps detailed in the bug, the behavior with this fix is as follows:
[sdasu@sdasu-thinkpadp1gen5 bin]$ ./openshift-install create cluster --dir ~/clusters/gcp-dev
INFO Consuming Common Manifests from target directory 
INFO Consuming Openshift Manifests from target directory 
INFO Consuming OpenShift Install (Manifests) from target directory 
INFO Consuming Worker Machines from target directory 
INFO Consuming Master Machines from target directory 
INFO Credentials loaded from file "/home/sdasu/.gcp/osServiceAccount.json" 
FATAL failed to fetch Terraform Variables: failed to generate asset "Terraform Variables": number of worker assets: 0 does not match requested workers: 3

File: pkg/asset/cluster/tfvars/tfvars.go [hunk 1]
```diff
@@ -501,11 +495,15 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
...
 		// 1. workers > 0, masters not schedulable, valid cluster
-		// 2. workers = 0, masters schedulable, valid compact cluster but currently unsupported
+		// 2. workers != compute replicas (possibly caused by worker manifests being manipulated after generation), masters schedulable, invalid cluster
 		// 3. workers = 0, masters not schedulable, invalid cluster
-		if len(workers) == 0 {
-			return errors.Errorf("compact clusters with 0 workers are not supported at this time")
+		if int64(len(workers)) != numWorkers {
+			return fmt.Errorf("number of worker assets: %d does not match requested workers: %d", len(workers), numWorkers)
+		}
+		if len(workers) == 0 && !mastersSchedulable {
```
Comment (barbacbd) [inline pkg/asset/cluster/tfvars/tfvars.go:503]:
```suggestion
		if len(workers) == 0 || !mastersSchedulable {
```
Should this be an or here?

File: pkg/asset/cluster/tfvars/tfvars.go [hunk 2]
```diff
@@ -219,6 +219,13 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 		return errors.Errorf("master slice cannot be empty")
 	}
 
+	numWorkers := int64(0)
+	for _, worker := range installConfig.Config.Compute {
+		if worker.Replicas != nil {
+			numWorkers += *worker.Replicas
+		}
```
Comment (r4f4) [inline pkg/asset/cluster/tfvars/tfvars.go:226]: There is a nice function for this in `k8s.io/utils/ptr`:
```suggestion
		numWorkers += ptr.Deref(worker.Replicas, 0)
```

File: pkg/asset/cluster/tfvars/tfvars.go [hunk 3]
```diff
@@ -501,11 +495,15 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 		}
 		// Based on the number of workers, we could have the following outcomes:
 		// 1. workers > 0, masters not schedulable, valid cluster
-		// 2. workers = 0, masters schedulable, valid compact cluster but currently unsupported
+		// 2. workers != compute replicas (possibly caused by worker manifests being manipulated after generation), masters schedulable, invalid cluster
 		// 3. workers = 0, masters not schedulable, invalid cluster
-		if len(workers) == 0 {
-			return errors.Errorf("compact clusters with 0 workers are not supported at this time")
+		if int64(len(workers)) != numWorkers {
```
Comment (r4f4) [inline pkg/asset/cluster/tfvars/tfvars.go:500]: We are comparing different things here:
* `len(workers)` == the number of worker machinesets
* `numWorkers` == the number of replicas in all worker machinesets.

Comment (r4f4) [reply pkg/asset/cluster/tfvars/tfvars.go:500]: As I suspected, GCP installs are broken:
```
level=info msg=Credentials loaded from environment variable "GOOGLE_CLOUD_KEYFILE_JSON", file "/var/run/secrets/ci.openshift.io/cluster-profile/gce.json"
level=fatal msg=failed to fetch Terraform Variables: failed to generate asset "Terraform Variables": number of worker assets: 4 does not match requested workers: 3
Installer exit with code 1
```
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/8370/pull-ci-openshift-installer-master-e2e-gcp-ovn/1788264305544663040

File: pkg/asset/cluster/tfvars/tfvars.go (hunk 2 above)
Comment (sadasu) [reply pkg/asset/cluster/tfvars/tfvars.go]: Done.

File: pkg/asset/cluster/tfvars/tfvars.go (hunk 1 above)
Comment (sadasu) [reply pkg/asset/cluster/tfvars/tfvars.go]: The condition is correct. I have updated the comment to hopefully make the current checks clearer.

File: pkg/asset/cluster/tfvars/tfvars.go (hunk 3 above)
Comment (sadasu) [reply pkg/asset/cluster/tfvars/tfvars.go]: I didn't realize that the number of worker manifests == number of AZs that the cluster is going to be installer in.

numWorkers which is the sum of compute replicas across all worker manifests is a better metric for our checks. Updated fix to incorporate this understanding.

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Pull request merged**